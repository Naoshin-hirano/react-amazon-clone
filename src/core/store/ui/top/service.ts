import { PayloadAction } from "@reduxjs/toolkit";
import { CartProduct, UITop } from "./type";

export const initialState = {
    productInCart: [
        {
            id: "",
            title: "",
            image: "",
            price: 0,
            rating: 0,
        },
    ],
    currentTotalAmount: 0,
    currentProductNum: 0,
};

export const reducers = {
    updateProductInCartAction: (
        _state: UITop,
        action: PayloadAction<CartProduct>
    ) => ({
        ..._state,
        productInCart: [..._state.productInCart, action.payload],
    }),
    updateCurrentTotalAmountAction: (
        _state: UITop,
        action: PayloadAction<number>
    ) => ({
        ..._state,
        currentTotalAmount: _state.currentTotalAmount + action.payload,
    }),
    updateCurrentProductNumAction: (_state: UITop) => ({
        ..._state,
        currentProductNum: _state.currentProductNum + 1,
    }),
};
