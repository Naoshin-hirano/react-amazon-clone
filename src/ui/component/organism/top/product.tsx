import React from "react";
import "../../../../design/component/atoms/top/product.css";

export const Product: React.FC<any> = ({
    id,
    title,
    price,
    rating,
    image,
    addToBasket,
}: any) => {
    return (
        <div className="product">
            <div className="product__info">
                <p>{title}</p>
                <p>{id}</p>
                <p className="product__price">
                    <small>¥</small>
                    <strong>{price}</strong>
                </p>
                <div className="product__rating">
                    {Array(rating)
                        .fill(undefined)
                        .map((_, key) => (
                            <p key={key}>🌟</p>
                        ))}
                </div>
            </div>
            <div className="product__rating"></div>

            <img src={image} alt="The Lean Startup book cover" />

            <button
                onClick={() =>
                    addToBasket({
                        id,
                        title,
                        image,
                        price,
                        rating,
                    })
                }
            >
                カートへ
            </button>
        </div>
    );
};
