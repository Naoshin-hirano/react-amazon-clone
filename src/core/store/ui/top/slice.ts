import { createSlice } from "@reduxjs/toolkit";
import { initialState, reducers } from "./service";

export const UITopSlice = createSlice({
    name: "ui/top",
    initialState,
    reducers,
});
