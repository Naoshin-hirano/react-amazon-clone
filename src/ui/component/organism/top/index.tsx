import React from "react";
import "../../../../design/component/atoms/top/index.css";
import { Product } from "./product";
import Carousel from "react-material-ui-carousel";
import Paper from "@material-ui/core/Paper";
import { images } from "./images";

export const Top: React.FC<any> = () => {
    return (
        <div className="home">
            <div className="home__container">
                <div className="home__image">
                    <Carousel
                        navButtonsAlwaysVisible={false}
                        indicators={false}
                        fullHeightHover={false}
                    >
                        {images.map((image, i) => (
                            <Image key={i} image={image} />
                        ))}
                    </Carousel>
                </div>

                <div className="home__rowContainer">
                    <div className="home__row">
                        <Product
                            id="12321341"
                            title="The Lean Startup: How Constant Innovation Creates Radically Successful Businesses Paperback"
                            price={1196}
                            rating={5}
                            image="https://images-na.ssl-images-amazon.com/images/I/51Zymoq7UnL._SX325_BO1,204,203,200_.jpg"
                        />
                        <Product
                            id="49538094"
                            title="Kenwood kMix Stand Mixer for Baking, Stylish Kitchen Mixer with K-beater, Dough Hook and Whisk, 5 Litre Glass Bowl"
                            price={239}
                            rating={4}
                            image="https://images-na.ssl-images-amazon.com/images/I/81O%2BGNdkzKL._AC_SX450_.jpg"
                        />
                    </div>

                    <div className="home__row">
                        <Product
                            id="4903850"
                            title="Samsung LC49RG90SSUXEN 49' Curved LED Gaming Monitor"
                            price={199}
                            rating={3}
                            image="https://images-na.ssl-images-amazon.com/images/I/71Swqqe7XAL._AC_SX466_.jpg"
                        />
                        <Product
                            id="23445930"
                            title="Amazon Echo (3rd generation) | Smart speaker with Alexa, Charcoal Fabric"
                            price={98}
                            rating={5}
                            image="https://media.very.co.uk/i/very/P6LTG_SQ1_0000000071_CHARCOAL_SLf?$300x400_retinamobilex2$"
                        />
                        <Product
                            id="3254354345"
                            title="New Apple iPad Pro (12.9-inch, Wi-Fi, 128GB) - Silver (4th Generation)"
                            price={598}
                            rating={4}
                            image="https://images-na.ssl-images-amazon.com/images/I/816ctt5WV5L._AC_SX385_.jpg"
                        />
                    </div>

                    <div className="home__row">
                        <Product
                            id="90829332"
                            title="Samsung LC49RG90SSUXEN 49' Curved LED Gaming Monitor - Super Ultra Wide Dual WQHD 5120 x 1440"
                            price={1094}
                            rating={4}
                            image="https://images-na.ssl-images-amazon.com/images/I/6125mFrzr6L._AC_SX355_.jpg"
                        />
                    </div>
                </div>
            </div>
        </div>
    );
};

const Image: React.FC<any> = ({ image }) => {
    return (
        <Paper>
            <img src={image.src} alt="amazon banner" />
        </Paper>
    );
};
