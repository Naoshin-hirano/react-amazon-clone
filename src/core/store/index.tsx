import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { UIProfileSlice } from "./ui/profile/slice";
import { UITopSlice } from "./ui/top/slice";

const appReducer = combineReducers({
    [UIProfileSlice.name]: UIProfileSlice.reducer,
    [UITopSlice.name]: UITopSlice.reducer,
});

const store = configureStore({
    reducer: appReducer,
});

const { dispatch } = store;
export type StateType = typeof store.getState;

export { store, dispatch };
