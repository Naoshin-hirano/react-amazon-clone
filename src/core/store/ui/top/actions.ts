import { UITopSlice } from "./slice";

export const UITopAction = UITopSlice.actions;
