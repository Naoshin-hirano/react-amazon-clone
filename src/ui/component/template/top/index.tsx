import React from "react";
import { Header } from "ui/component/organism/common/header";
import { Top } from "ui/component/organism/top";

export const Template: React.FC<any> = (props) => {
    return (
        <div>
            <Header />
            <Top {...props} />
        </div>
    );
};
