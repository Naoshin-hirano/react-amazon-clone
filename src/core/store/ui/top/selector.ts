import { createSelector } from "@reduxjs/toolkit";
import { StateType } from "core/store";
import { UITop } from "./type";

export const UITopSelector = createSelector(
    (state: ReturnType<StateType>) => state["ui/top"],
    (state: UITop) => state
);
