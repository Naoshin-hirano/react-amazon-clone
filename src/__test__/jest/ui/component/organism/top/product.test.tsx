import { fireEvent, render } from "@testing-library/react";
import React from "react";
import { act } from "react-dom/test-utils";
import { Product } from "ui/component/organism/top/product";
import * as Store from "core/store";

const mockHistoryPush = jest.fn();
const addToBasket = jest.fn();
const dispatchSpy = jest.spyOn(Store, "dispatch");

jest.mock("@tanstack/react-location", () => {
    return {
        ...jest.requireActual("@tanstack/react-location"),
        useLocation: () => ({
            history: {
                push: mockHistoryPush,
            },
        }),
    };
});

describe("Header Unit Test", () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("Topコンポーネント snapshotテスト", async () => {
        const { asFragment } = render(
            <Product
                id="12321341"
                title="The Lean Startup: How Constant Innovation Creates Radically Successful Businesses Paperback"
                price={1196}
                rating={5}
                image="https://images-na.ssl-images-amazon.com/images/I/51Zymoq7UnL._SX325_BO1,204,203,200_.jpg"
                addToBasket={addToBasket}
            />
        );
        expect(asFragment()).toMatchSnapshot("Product");
    });

    it("カートへボタン押下", async () => {
        const wrapper = render(
            <Product
                id="12321341"
                title="The Lean Startup: How Constant Innovation Creates Radically Successful Businesses Paperback"
                price={1196}
                rating={5}
                image="https://images-na.ssl-images-amazon.com/images/I/51Zymoq7UnL._SX325_BO1,204,203,200_.jpg"
                addToBasket={addToBasket}
            />
        );
        await act(async () => {
            const buttonName = wrapper.getByText("カートへ");
            await fireEvent.click(buttonName);
        });
        expect(dispatchSpy).toHaveBeenCalledTimes(0);
        expect(mockHistoryPush).toHaveBeenCalledTimes(0);
    });
});
