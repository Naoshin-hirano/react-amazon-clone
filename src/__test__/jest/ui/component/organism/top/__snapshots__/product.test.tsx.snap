// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Unit Test Topコンポーネント snapshotテスト: Product 1`] = `
<DocumentFragment>
  <div
    class="product"
  >
    <div
      class="product__info"
    >
      <p>
        The Lean Startup: How Constant Innovation Creates Radically Successful Businesses Paperback
      </p>
      <p>
        12321341
      </p>
      <p
        class="product__price"
      >
        <small>
          ¥
        </small>
        <strong>
          1196
        </strong>
      </p>
      <div
        class="product__rating"
      >
        <p>
          🌟
        </p>
        <p>
          🌟
        </p>
        <p>
          🌟
        </p>
        <p>
          🌟
        </p>
        <p>
          🌟
        </p>
      </div>
    </div>
    <div
      class="product__rating"
    />
    <img
      alt="The Lean Startup book cover"
      src="https://images-na.ssl-images-amazon.com/images/I/51Zymoq7UnL._SX325_BO1,204,203,200_.jpg"
    />
    <button>
      カートへ
    </button>
  </div>
</DocumentFragment>
`;
